docs/rfc/rfc-016-node-architecture.md

```diff
@@ -0,0 +1,78 @@
+# RFC 016: Node Architecture
+
+## Changelog
+
+- April 8, 2022: Initial draft (@cmwaters)
+
+## Abstract
+
+The `node` package is the entry point into the Tendermint codebase, used both by the command line and programatically to create the nodes that make up a network. The package has suffered the most from the evolution of the codebase, becoming bloated as developers clipped on their bits of code here and there to get whatever feature they wanted working. 
+
+The decisions made at the node level have the biggest impact to simplifying the protocols within them, unlocking better internal designs and making Tendermint more intuitive to use and easier to understand from the outside. Work, in minor increments, has already begun on this section of the codebase. This document exists to spark forth the necessary discourse in a few related areas that will help the team to converge on the long term makeup of the node.
+
+## Background
+
+> To be written. This will mostly entail the less controversial work that Sam has been doing to clean up the node. 
+
+## Discussion
+
+The following is a list of points of discussion around the architecture of the node:
+
+### Dependency Tree
+
+The node object is currently stuffed with every component that possibly exists within Tendermint. In the constructor, all objects are built and interlaid with one another in some awkward dance. My guiding principle is that the node should only be made up of the components that it wants to have direct control of throughout its life. The node is a service which currently has the purpose of starting other services up in a particular order and stopping them all when commanded to do so. However, there are many services which are direct dependents i.e. the mempool and evidence services should only should be working when the consensus service is running. I propose to form more of a hierarchical structure of dependents which forces us to be clear about the relations that one component has to the other (as opposed to having a free for all with information passing from all components to one another). More concretely, I propose the following dependency tree:
+
+- Node (also the orchestrator which I cover more in the next section)
+  - Networking/P2P (service to allow communication across nodes via channels)
+    - PEX reactor (or whatever this turns into post libP2P)
+  - Consensus (network wide state advancing mechanism)
+    - Block Executor (the central component used to advance state via blocks)
+    - Mempool (gossiping of transactions. To be used by the proposer)
+    - Evidence (gossiping of evidence. To be used by the proposer)
+  - Blocksync (local node state advancing mechanism)
+    - Block Exector
+  - Statesync (local node state advancing mechanism)
+    - Block Executor
+
+Note that I've intentionally omitted any mention of the RPC because it's a challenging issue that deserves it's own section.
```

Comment:
Unfortunately, a lot of the existing complexity is tied up in the cross-cutting concerns like this (RPC) and the need of other components to pass information to each other. Metrics and logging are other examples, although those both only require one-way communication (output only) and thus easier.

A hierarchy of control is appealing, because it's easy to understand which components are responsible for which others. To the extent that configuration can flow down the tree, this is pretty easy to set up. The challenge, however, is that lifecycle control is not the only consideration.

We currently have an event bus shared by the system that could address a lot of these other concerns. Right now it is not providing much value, in that it's primarily used by the indexer and to detect state transitions in the unit tests.  In my background research for working on indexing and the RPC events API, it seems clear that the original aspirations were larger: To have the event bus be more of a "nervous system" for the various components of the node.

I think if we really want to tease apart these concerns, we have to pick apart the RPC knot. One possible approach we could take would be along these lines:

- The RPC system is started up before anything else in the node comes up, along with metric, profiling, etc. This ensures the entire lifecycle of the remaining components is accessible to diagnostic tools
- Initially, the RPC module is "empty", meaning it has only global methods (e.g., health check).
- This module is then plumbed down the control tree along with the configuration settings. Each component plumbs it down into its subordinates.
- Each component can "register" methods with the RPC module, which adds routes to the visible service. When a component exits, these methods can be unregistered.
- When a method is registered, its handler (provided by the component that did the registration) is responsible for encapsulating all the state the handler needs to serve its requests.

There are a couple key points here:

1. The RPC subsystem would no longer need a static pointer to every other component.
2. We would have a clean way to control the visible surface area of the RPC service.

A similar approach could be taken to plumbing an event bus.  I think a lot of the cases where we currently have weird initialization order issues could be addressed by having a policy that components only communicate directly with their immediate superior and their subordinates, and all other intra-component communication happens on the event bus.

That's different from how we do it now, in that most things wind up wired together in a pointer graph with method calls. But I think we could strangler-fig our way from a pointer graph to a shared bus without too much pain.

Reply:
I like the idea of turning this all on it's head and saying "let's try and detangle the RPC system and use that as a way of disconnecting the underlying components.

I'm a little worried, in practice that this will mean passing a "register RPC method/hook) to all of the reactors (or potentially lashed to the context that you send to on-start). This assumes that the endpoints don't actually depend on more than one component, which isn't an assumption I'd make, but is easy enough to check,) but I think this is likely to be fine. 

Similarly, I'm a little worried about using the eventbus plumbing for intra-node communication. In part because I feel like we're sort of continually being burned by it so my baseline level of trust is not high (this is fixable), and because I think it effectively exposes communication that is probably fine to expose but I'd just as soon not have to think about that.

Reply:
> This assumes that the endpoints don't actually depend on more than one component, which isn't an assumption I'd make, but is easy enough to check,) but I think this is likely to be fine.

I'm guessing that, at the moment, they are somewhat interdependent, although truthfully I'm not certain. I would say that endpoints that perform write operations should, as best as we can, be modeled as relating to only one component/module. If an RPC endpoint that, say, submits transactions, cannot be wholly implemented using just the `mempool` code, that's a bit of a code smell. 

Either way, I think it's a reasonable north star that will leave us with likely a few edge cases that we'll have to sort out either by 1. updating our package boundaries or 2. updating our RPC endpoints both of which are not particularly dramatic.

Reply:
> I'm a little worried, in practice that this will mean passing a "register RPC method/hook) to all of the reactors (or potentially lashed to the context that you send to on-start).

Yes, that's exactly what I'm proposing. My mental model here is that the interface for all components will accept a package of common settings. That would include the Config value, which everything needs, plus (in this formulation) a pointer to the RPC component and the event bus. Those three are the only pieces "everything" needs, I believe.

> This assumes that the endpoints don't actually depend on more than one component, which isn't an assumption I'd make, but is easy enough to check,) but I think this is likely to be fine.

Yes, that's right, but I think that's a safe assumption: Right now the RPC system accesses lots of different components to serve different endpoints, but broadly each method only uses a subset. As long as the subset for a method has a common ancestor in the control tree, that ancestor can be responsible for adding/removing the endpoint hooks.

That's not to say we _have_ to do it this way, only that it's feasible.

> Similarly, I'm a little worried about using the eventbus plumbing for intra-node communication.

Yes, and the event bus would be the other "shared" component: config + RPC + bus.

Reply:
> Yes, and the event bus would be the other "shared" component: config + RPC + bus.

Well and just about everything needs the p2p router too.

Reply:
> Well and just about everything needs the p2p router too.

That makes sense. I think the key property all these things share is that they are dependency roots (a.k.a. "infrastructure"): They do not(*) depend on anything else and can be started up at basically any time.

The (*) here is to acknowledge that RPC right now _does_ depend on other things. But it doesn't need to. Other things ideally should use it as a substrate, as with P2P routing.

Reply:
I've tried to incorporate this conversation into the RFC document itself